File: paleo_workbench/agent/planner.py
"""DAG Task Graph Planner for Paleo AI GIS Harness.

Generates dependency graphs and schedules multi-agent execution steps.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Any

from paleo_workbench.agent.intent import ParsedIntent, TaskDomain


class TaskStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    SKIPPED = "skipped"


@dataclass
class TaskNode:
    id: str
    agent_name: str
    action: str
    description: str
    dependencies: list[str] = field(default_factory=list)
    parameters: dict[str, Any] = field(default_factory=dict)
    status: TaskStatus = TaskStatus.PENDING
    result: Any = None
    error: str | None = None


@dataclass
class TaskGraph:
    nodes: dict[str, TaskNode] = field(default_factory=dict)

    def add_node(self, node: TaskNode) -> None:
        self.nodes[node.id] = node

    def get_executable_nodes(self) -> list[TaskNode]:
        """Return nodes that are PENDING and whose dependencies are all COMPLETED."""
        ready = []
        for node in self.nodes.values():
            if node.status != TaskStatus.PENDING:
                continue
            deps_ok = all(
                self.nodes[dep_id].status == TaskStatus.COMPLETED
                for dep_id in node.dependencies
                if dep_id in self.nodes
            )
            if deps_ok:
                ready.append(node)
        return ready

    def is_finished(self) -> bool:
        return all(node.status in {TaskStatus.COMPLETED, TaskStatus.FAILED, TaskStatus.SKIPPED} for node in self.nodes.values())

    def has_failures(self) -> bool:
        return any(node.status == TaskStatus.FAILED for node in self.nodes.values())


class TaskPlanner:
    """Intelligent DAG Planner that breaks domain intent into multi-agent task steps."""

    def create_plan(self, intent: ParsedIntent) -> TaskGraph:
        graph = TaskGraph()

        # Step 1: Data Discovery & Inspection (Always First)
        graph.add_node(
            TaskNode(
                id="task_data_discover",
                agent_name="data_agent",
                action="discover_and_validate",
                description="Discover and validate catalog data assets and schemas",
                dependencies=[],
                parameters=intent.parameters,
            )
        )

        # Step 2: Domain-specific Processing
        if intent.primary_domain in {TaskDomain.WELL_LOGGING, TaskDomain.SINGLE_FACTOR_MAPPING, TaskDomain.PALEOMAP_COMPILATION}:
            graph.add_node(
                TaskNode(
                    id="task_well_process",
                    agent_name="well_agent",
                    action="process_well_tops_and_curves",
                    description="Extract formation tops and align well log curves",
                    dependencies=["task_data_discover"],
                    parameters=intent.parameters,
                )
            )

        if intent.primary_domain in {TaskDomain.SEISMIC_INTERPRETATION, TaskDomain.PALEOMAP_COMPILATION}:
            graph.add_node(
                TaskNode(
                    id="task_seismic_process",
                    agent_name="seismic_agent",
                    action="extract_slices_and_attributes",
                    description="Extract 3D seismic slices and coherence volumes",
                    dependencies=["task_data_discover"],
                    parameters=intent.parameters,
                )
            )

        # Step 3: Spatial & GIS Analysis
        graph.add_node(
            TaskNode(
                id="task_gis_spatial",
                agent_name="gis_agent",
                action="analyze_spatial_constraints",
                description="Extract fault barriers and validate boundary topologies",
                dependencies=["task_data_discover"],
                parameters=intent.parameters,
            )
        )

        # Step 4: Single-Factor / Paleomap Cartography
        carto_deps = ["task_gis_spatial"]
        if "task_well_process" in graph.nodes:
            carto_deps.append("task_well_process")
        if "task_seismic_process" in graph.nodes:
            carto_deps.append("task_seismic_process")

        graph.add_node(
            TaskNode(
                id="task_carto_generate",
                agent_name="carto_agent",
                action="generate_factor_surface",
                description="Perform barrier-constrained anisotropic IDW and contouring",
                dependencies=carto_deps,
                parameters=intent.parameters,
            )
        )

        # Step 5: Map Composition & Visualization
        graph.add_node(
            TaskNode(
                id="task_viz_compose",
                agent_name="viz_agent",
                action="compose_map_layout",
                description="Compose standardized geological map layout with legends and graticules",
                dependencies=["task_carto_generate"],
                parameters=intent.parameters,
            )
        )

        # Step 6: QA / QC Verification
        graph.add_node(
            TaskNode(
                id="task_qa_audit",
                agent_name="qa_agent",
                action="audit_and_verify",
                description="Execute topological, boundary, and statistical QC audit",
                dependencies=["task_viz_compose"],
                parameters=intent.parameters,
            )
        )

        # Step 7: Final Delivery & Result Reporting
        graph.add_node(
            TaskNode(
                id="task_result_delivery",
                agent_name="result_agent",
                action="package_and_report",
                description="Assemble execution report, lineage metadata, and deliverables",
                dependencies=["task_qa_audit"],
                parameters=intent.parameters,
            )
        )

        return graph


task_planner = TaskPlanner()

File: libs/closure_agent/src/planner.cpp
// planner.cpp — TaskPlanner port (agent/planner.py): the fixed skeleton,
// domain-conditioned well/seismic steps, and cartography fan-in.
#include <pwb/closure_agent/planner.hpp>

namespace pwb::closure_agent {

std::string to_string(TaskStatus status) {
    switch (status) {
        case TaskStatus::Pending: return "pending";
        case TaskStatus::Running: return "running";
        case TaskStatus::Completed: return "completed";
        case TaskStatus::Failed: return "failed";
        case TaskStatus::Skipped: return "skipped";
    }
    return "pending";
}

std::optional<TaskStatus> task_status_from_string(const std::string& value) {
    if (value == "pending") return TaskStatus::Pending;
    if (value == "running") return TaskStatus::Running;
    if (value == "completed") return TaskStatus::Completed;
    if (value == "failed") return TaskStatus::Failed;
    if (value == "skipped") return TaskStatus::Skipped;
    return std::nullopt;
}

void TaskGraph::add_node(TaskNode node) {
    nodes.push_back(std::move(node));
}

TaskNode* TaskGraph::find(const std::string& id) {
    for (auto& node : nodes) {
        if (node.id == id) return &node;
    }
    return nullptr;
}

const TaskNode* TaskGraph::find(const std::string& id) const {
    for (const auto& node : nodes) {
        if (node.id == id) return &node;
    }
    return nullptr;
}

std::vector<TaskNode*> TaskGraph::executable_nodes() {
    std::vector<TaskNode*> ready;
    for (auto& node : nodes) {
        if (node.status != TaskStatus::Pending) continue;
        bool deps_ok = true;
        for (const auto& dep_id : node.dependencies) {
            const TaskNode* dep = find(dep_id);
            if (dep != nullptr && dep->status != TaskStatus::Completed) {
                deps_ok = false;
                break;
            }
        }
        if (deps_ok) ready.push_back(&node);
    }
    return ready;
}

bool TaskGraph::is_finished() const {
    for (const auto& node : nodes) {
        if (node.status == TaskStatus::Pending || node.status == TaskStatus::Running) {
            return false;
        }
    }
    return true;
}

bool TaskGraph::has_failures() const {
    for (const auto& node : nodes) {
        if (node.status == TaskStatus::Failed) return true;
    }
    return false;
}

TaskGraph TaskPlanner::create_plan(const ParsedIntent& intent) const {
    TaskGraph graph;
    const Json& parameters = intent.parameters;

    TaskNode discover;
    discover.id = "task_data_discover";
    discover.agent_name = "data_agent";
    discover.action = "discover_and_validate";
    discover.description = "Discover and validate catalog data assets and schemas";
    discover.parameters = parameters;
    graph.add_node(std::move(discover));

    const bool well_needed =
        intent.primary_domain == TaskDomain::WellLogging ||
        intent.primary_domain == TaskDomain::SingleFactorMapping ||
        intent.primary_domain == TaskDomain::PaleomapCompilation;
    if (well_needed) {
        TaskNode well;
        well.id = "task_well_process";
        well.agent_name = "well_agent";
        well.action = "process_well_tops_and_curves";
        well.description = "Extract formation tops and align well log curves";
        well.dependencies = {"task_data_discover"};
        well.parameters = parameters;
        graph.add_node(std::move(well));
    }

    const bool seismic_needed =
        intent.primary_domain == TaskDomain::SeismicInterpretation ||
        intent.primary_domain == TaskDomain::PaleomapCompilation;
    if (seismic_needed) {
        TaskNode seismic;
        seismic.id = "task_seismic_process";
        seismic.agent_name = "seismic_agent";
        seismic.action = "extract_slices_and_attributes";
        seismic.description = "Extract 3D seismic slices and coherence volumes";
        seismic.dependencies = {"task_data_discover"};
        seismic.parameters = parameters;
        graph.add_node(std::move(seismic));
    }

    TaskNode gis;
    gis.id = "task_gis_spatial";
    gis.agent_name = "gis_agent";
    gis.action = "analyze_spatial_constraints";
    gis.description = "Extract fault barriers and validate boundary topologies";
    gis.dependencies = {"task_data_discover"};
    gis.parameters = parameters;
    graph.add_node(std::move(gis));

    std::vector<std::string> carto_deps = {"task_gis_spatial"};
    if (graph.find("task_well_process") != nullptr) {
        carto_deps.push_back("task_well_process");
    }
    if (graph.find("task_seismic_process") != nullptr) {
        carto_deps.push_back("task_seismic_process");
    }
    TaskNode carto;
    carto.id = "task_carto_generate";
    carto.agent_name = "carto_agent";
    carto.action = "generate_factor_surface";
    carto.description = "Perform barrier-constrained anisotropic IDW and contouring";
    carto.dependencies = std::move(carto_deps);
    carto.parameters = parameters;
    graph.add_node(std::move(carto));

    TaskNode viz;
    viz.id = "task_viz_compose";
    viz.agent_name = "viz_agent";
    viz.action = "compose_map_layout";
    viz.description =
        "Compose standardized geological map layout with legends and graticules";
    viz.dependencies = {"task_carto_generate"};
    viz.parameters = parameters;
    graph.add_node(std::move(viz));

    TaskNode qa;
    qa.id = "task_qa_audit";
    qa.agent_name = "qa_agent";
    qa.action = "audit_and_verify";
    qa.description = "Execute topological, boundary, and statistical QC audit";
    qa.dependencies = {"task_viz_compose"};
    qa.parameters = parameters;
    graph.add_node(std::move(qa));

    TaskNode delivery;
    delivery.id = "task_result_delivery";
    delivery.agent_name = "result_agent";
    delivery.action = "package_and_report";
    delivery.description =
        "Assemble execution report, lineage metadata, and deliverables";
    delivery.dependencies = {"task_qa_audit"};
    delivery.parameters = parameters;
    graph.add_node(std::move(delivery));

    return graph;
}

}  // namespace pwb::closure_agent
